addons/mail/controllers/google_translate.py

```diff
@@ -0,0 +1,50 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request, route, Controller
+
+import requests
+
+
+class GoogleTranslateController(Controller):
+    @route("/mail/message/translate", type="json", auth="user")
+    def translate(self, message_id):
+        message = request.env["mail.message"].browse(message_id).exists()
+        message.ensure_one()._validate_access_for_current_persona("read")
+        MessageTranslation = request.env["mail.message.translation"].sudo()
+        user_lang = request.env.user.lang.split("_")[0]
+        translation = MessageTranslation.search([("message_id", "=", message_id), ("target_lang", "=", user_lang)])
+        if not translation:
+            try:
+                body = str(message.body)
+                source = self._detect_source(body, message_id)
+                text = self._get_translation(body, source)
+                translation = MessageTranslation.create(
+                    {
+                        "message_id": message_id,
+                        "body": text,
+                        "source_lang": source,
+                    }
+                )
+            except requests.exceptions.HTTPError as err:
+                return {"error": err.response.json()["error"]["message"]}
+        return {"source_lang": translation.format_source(), "body": translation.body}
+
+    def _detect_source(self, body, message_id):
+        MessageTranslation = request.env["mail.message.translation"].sudo()
+        translation = MessageTranslation.search([("message_id", "=", message_id)], limit=1)
+        if translation:
+            return translation.source_lang
+        response = self._post("/detect", {"q": body})
+        return response.json()["data"]["detections"][0][0]["language"]
+
+    def _get_translation(self, body, source):
+        user_lang = request.env.user.lang.split("_")[0]
+        response = self._post("", {"q": body, "target": user_lang, "source": source})
+        return response.json()["data"]["translations"][0]["translatedText"]
+
+    def _post(self, endpoint, data):
+        api_key = request.env["ir.config_parameter"].sudo().get_param("mail.google_translate_api_key")
+        url = f"https://translation.googleapis.com/language/translate/v2{endpoint}/?key={api_key}"
```

Comment:
minor : it's weird to have to pass the `/` in the endpoint string, I would not (and did not in the gif picker PR).

Reply:
In my opinion, it was the simplest way to manage empty path. 
Maybe this part should be refactored to use werkzeug instead...

Reply:
> In my opinion, it was the simplest way to manage empty path.

Well it's not that good to have `_post("", data)`.
One way to do it would be to leverage named param instead of positional param with this signature:
`def _post(self, endpoint="", data=None):`
It doesn't help for the `/` but at least you don't pass an empty string.